### ISSUE TITLE
oidc: No permissions to list pods (AWS Cognito)

### ISSUE THREAD

**Issue by javlonsodikov** [kinvolk/headlamp#393]
## Description

oidc: No permissions to list pods.

## Impact

Cannot use headlamp after OIDC login

## Environment and steps to reproduce

1. **Set-up**: headlamp, configure oidc with AWS cognito  
2. **Task**: Login to the headlamp 
3. **Action(s)**: 
  a. open main page
  b. press "Sign In"
  c. type OIDC login password, login

4. **Error**: 
     page redirects to the main headlamp page but "No permissions to list pods." errors messages are shown. 
## Expected behavior

Expected to see cluster details 

## Additional information

What else should be configured?

**Comment by joaquimrocha** [comment]
Hi @javlonsodikov , does Headlamp keep redirecting you to the login, or does it feel like logging in works but it's the listing that fails?
If that's the case, please double check you have the right role to list pods, as Headlamp will check those permissions in order to determine if any cluster resource can be accessed/updated/deleted/etc.

**Comment by javlonsodikov** [comment]
I would say both: after logging in I see following: 
![image](https://user-images.githubusercontent.com/5047093/149385813-b3db3dc0-5896-4e6d-a338-5e8bc415eb16.png)

Whenever I click any other sections in the UI its redirects to login page

also in the chrome console I can see 
```
{kind: "Status", apiVersion: "v1", metadata: {}, status: "Failure",…}
apiVersion: "v1"
code: 403
details: {group: "metrics.k8s.io", kind: "nodes"}
kind: "Status"
message: "nodes.metrics.k8s.io is forbidden: User \"system:anonymous\" cannot list resource \"nodes\" in API group \"metrics.k8s.io\" at the cluster scope"
metadata: {}
reason: "Forbidden"
status: "Failure"
```

**Comment by javlonsodikov** [comment]
here is the authorization Bearer 
```
{
  "sub": "fb6cf411-aff6-4630-964f-899a03a94b45",
  "cognito:groups": [
    "cluster-admin",
  ],
  "iss": "https://cognito-idp.us-east-1.amazonaws.com/<id>",
  "version": 2,
  "client_id": "<client-id>",
  "origin_jti": "<masted_jti>",
  "token_use": "access",
  "scope": "phone openid profile email",
  "auth_time": 1642158722,
  "exp": 1642162322,
  "iat": 1642158722,
  "jti": "<jti>",
  "username": "fb6cf411-aff6-4630-964f-899a03a94b45"
}
```

**Comment by joaquimrocha** [comment]
@javlonsodikov , Sorry we've been busy. From your report, it looks like the role your user has doesn't allow to view any resources. So I would say this is a matter of making sure your token includes the right role (cluster role view at least, AFAIU).

Nevertheless, we will try to investigate it with incognito next week so we discard any issue with our auth system. We will keep you posted.

**Comment by javlonsodikov** [comment]
@joaquimrocha, thanks for the reply. Also would be good to have proper kubectl command to create a role

**Comment by ashu8912** [comment]
@javlonsodikov I tried setting up headlamp with cognito as the oidc provider, and it worked for me
I am attaching the steps that I Followed maybe you can reverify and see if you missed something
First I tried setting up the user pool that I created in cognito for my cluster and that I achieved with 
```
kubectl config set-credentials USER_NAME \
   --auth-provider=oidc \
   --auth-provider-arg=idp-issuer-url= https://cognito-ikk.us-east-1.amazonaws.com/<user_pool_id>\
   --auth-provider-arg=client-id= <client_id_from_cognito> \
   --auth-provider-arg=client-secret= <client_secret_from_cognito>
```

Now if you are running headlamp locally by running the frontend and backend separately then the callback URL in cognito needs to be ```http://localhost:4466/oidc-callback```

If you followed a different approach please do reach out I will try it as well.
Thanks.

**Comment by javlonsodikov** [comment]
Thanks, @ashu8912 for the quick reply, much appreciated 
Could you please update the documentation here https://kinvolk.github.io/headlamp/docs/latest/installation/in-cluster/oidc/  as well?

**Comment by joaquimrocha** [comment]
> @joaquimrocha, thanks for the reply. Also would be good to have proper kubectl command to create a role

@javlonsodikov , I am curious to understand whether the solution @ashu8912 suggested works for you.

If there are still issues and you need to access headlamp using a token, please check https://kinvolk.github.io/headlamp/docs/latest/installation/#create-a-service-account-token .

**Comment by javlonsodikov** [comment]
@joaquimrocha proposed method didn't work. 
how to create oidc a user in kubectl? Maybe I am missing something. 

however, headlamp works perfectly with ServiceAccount token. But OIDC required by specs here

**Comment by ashu8912** [comment]
@javlonsodikov Thanks for reaching out, I think I misunderstood your setup, I tested it with a Kube config that has clusters that allow resource access to be done via oidc logins and that's why I specified the commands where you can configure oidc authentication for your cluster, But I think you are trying to run it in-cluster which I still haven't tested but will do it now.
Will get back to you soon with the results.

**Comment by ashu8912** [comment]
@javlonsodikov I tried headlamp with oidc on AWS with Cognito and it worked for me.
Can you please recheck if oidc is configured correctly for your cluster?

**Comment by javlonsodikov** [comment]
Cloud you please send me server-side and client-side configs?

**Comment by ashu8912** [comment]
@javlonsodikov The cognito thing worked for me when directly accessing my cluster with the correct user authorization and I assumed it would work with headlamp too because we had something similar to this done for an old cluster of ours, I tried the configuration with headlamp and it seems to be broken on our side. 
Thanks For reporting this.
**Note:** oidc configuration on an in-cluster deployment for headlamp is not working. Sorry for the wrong assumption I made, We are looking into the cause of this and will get back to you soon with the fix.

**Comment by javlonsodikov** [comment]
@ashu8912 please provide client and server settings. 
Also how commands to create user in k8s

**Comment by javlonsodikov** [comment]
Hello @ashu8912 
any progress on the issue?

**Comment by ashu8912** [comment]
Hello, @javlonsodikov Sorry this took some time but i think we finally found the fix,
I have created a PR with the fix.
you can try the PR or wait for it to land in main branch.
alternatively you can temporarily  try the image ```docker.io/ashu8912/headlamp-test``` in your development.yaml configuration to test this.

If you still run into problems please reach out, I have already tested it with AWS and cognito now so I think i can help you with the configurations.

**Comment by javlonsodikov** [comment]
I am still having the same issues. 
Please provide configuration details in cognito and in the client-side.

**Comment by ashu8912** [comment]
@javlonsodikov just one question what is your cluster environment??
also setting up

> here is the authorization Bearer
> 
> ```
> {
>   "sub": "fb6cf411-aff6-4630-964f-899a03a94b45",
>   "cognito:groups": [
>     "cluster-admin",
>   ],
>   "iss": "https://cognito-ikk.us-east-1.amazonaws.com/<id>",
>   "version": 2,
>   "client_id": "<client-id>",
>   "origin_jti": "<masted_jti>",
>   "token_use": "access",
>   "scope": "phone openid profile email",
>   "auth_time": 1642158722,
>   "exp": 1642162322,
>   "iat": 1642158722,
>   "jti": "<jti>",
>   "username": "fb6cf411-aff6-4630-964f-899a03a94b45"
> }
> ```

from here i see that you have a cluster-admin group in the cognito pool do you have a cluster role and cluster role binding as well pointing to this group??

**Comment by javlonsodikov** [comment]
the group was created for testing. its not working with/without group

**Comment by ashu8912** [comment]
Hey @javlonsodikov what configuration do you exactly need??

**Comment by ashu8912** [comment]
I can share with you the instructions that I followed specifically for an eks cluster that has headlamp deployed in-cluster and has congnito as the authentication provider.
I followed this guide to setup a cluster with cognito as the oidc provider 
https://aws.amazon.com/blogs/containers/introducing-oidc-identity-provider-authentication-amazon-eks/

**Comment by javlonsodikov** [comment]
for Cognito side, I will use this guide. 

Could you please provide configuration for EKS user creation? Maybe its obvious for you but I want to make sure that I am not missing anything.

**Comment by ashu8912** [comment]
So on EKS you just have to point your authentication provider to cognito pool
<img width="811" alt="Screenshot 2022-03-03 at 12 12 39 PM" src="https://user-images.githubusercontent.com/30126128/156511011-b4bbc24e-9da1-4282-9efd-de3a400c21c0.png">

**Comment by ashu8912** [comment]
@javlonsodikov we have a headlamp channel on kubernetes workspace if you would like to get it resolved over a call we can also do that there

**Comment by ashu8912** [comment]
This issue is resolved by #437
